### PR TITLE
Use FlashInfer tinygemm for GPT-OSS MoE router on SM90+

### PR DESCRIPTION
**Pull request by mmangkad**
## Summary

FlashInfer 0.6.6 (actually 0.6.5) added `tinygemm_bf16`, a faster kernel for small GEMMs. This applies it to the GPT-OSS MoE router on SM90+.

## Accuracy Test (GPQA)

Before:

```
[{'eval_name': 'gpqa', 'model_name': 'openai__gpt-oss-20b-low_temp1.0_20260317_060903', 'metric': 0.5625}]
```

After:

```
[{'eval_name': 'gpqa', 'model_name': 'openai__gpt-oss-20b-low_temp1.0_20260317_062812', 'metric': 0.5593434343434344}]
```

## Benchmarking and Profiling

Before:

```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 1
Successful requests:                     10
Benchmark duration (s):                  8.84
Total input tokens:                      2517
Total input text tokens:                 2517
Total generated tokens:                  2684
Total generated tokens (retokenized):    2576
Request throughput (req/s):              1.13
Input token throughput (tok/s):          284.65
Output token throughput (tok/s):         303.53
Peak output token throughput (tok/s):    307.00
Peak concurrent requests:                3
Total token throughput (tok/s):          588.18
Concurrency:                             1.00
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   882.95
Median E2E Latency (ms):                 847.49
P90 E2E Latency (ms):                    1527.18
P99 E2E Latency (ms):                    1654.68
---------------Time to First Token----------------
Mean TTFT (ms):                          12.41
Median TTFT (ms):                        12.05
P99 TTFT (ms):                           15.16
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          3.25
Median TPOT (ms):                        3.25
P99 TPOT (ms):                           3.26
---------------Inter-Token Latency----------------
Mean ITL (ms):                           3.26
Median ITL (ms):                         3.25
P95 ITL (ms):                            3.50
P99 ITL (ms):                            3.98
Max ITL (ms):                            8.07
==================================================
```

Screenshot 2026-03-17 143543

After:

```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 1
Successful requests:                     10
Benchmark duration (s):                  8.61
Total input tokens:                      2517
Total input text tokens:                 2517
Total generated tokens:                  2684
Total generated tokens (retokenized):    2521
Request throughput (req/s):              1.16
Input token throughput (tok/s):          292.39
Output token throughput (tok/s):         311.79
Peak output token throughput (tok/s):    316.00
Peak concurrent requests:                3
Total token throughput (tok/s):          604.19
Concurrency:                             1.00
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   859.66
Median E2E Latency (ms):                 826.02
P90 E2E Latency (ms):                    1486.07
P99 E2E Latency (ms):                    1610.24
---------------Time to First Token----------------
Mean TTFT (ms):                          12.22
Median TTFT (ms):                        11.53
P99 TTFT (ms):                           16.45
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          3.17
Median TPOT (ms):                        3.17
P99 TPOT (ms):                           3.18
---------------Inter-Token Latency----------------
Mean ITL (ms):                           3.17
Median ITL (ms):                         3.17
P95 ITL (ms):                            3.59
P99 ITL (ms):                            5.68
Max ITL (ms):                            9.44
==================================================
```

Screenshot 2026-03-17 143948

### CODE REVIEW

File: python/sglang/srt/models/gpt_oss.py [hunk 1]
```diff
@@ -97,6 +98,26 @@ def get_attention_sliding_window_size(config):
     return config.sliding_window - 1
 
 
+class GptOssRouterLinear(ReplicatedLinear):
```
Comment (zminglei) [inline python/sglang/srt/models/gpt_oss.py]: QQ, could the tinygemm benefit other MoE models as well? Or what makes gpt-oss special here and we're only applying it for gpt-oss?

Comment (mmangkad) [reply python/sglang/srt/models/gpt_oss.py]: @zminglei `tinygemm2` was originally integrated upstream in trtllm for gpt-oss, but it could be worth trying on other BF16 MoE routers in a follow-up too

File: python/sglang/srt/models/gpt_oss.py [hunk 2]
```diff
@@ -22,6 +22,7 @@
 from typing import Any, Dict, List, Optional, Tuple, Union
 
 import torch
+from flashinfer.gemm import tinygemm_bf16
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/models/gpt_oss.py]: Shall we guard this with `if is_cuda`?

File: python/sglang/srt/models/gpt_oss.py [hunk 3]
```diff
@@ -97,6 +98,39 @@ def get_attention_sliding_window_size(config):
...
 
+class GptOssRouterLinear(ReplicatedLinear):
+    """ReplicatedLinear with a FlashInfer tinygemm BF16 fast path."""
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        if (
+            x.ndim == 2
+            and x.is_cuda
+            and torch.cuda.get_device_capability(x.device)[0] >= 9
+            and x.shape[0] <= 128
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/models/gpt_oss.py]: cache these conditions in another variable, then we do not need to check all these every time in the hot path, especially conditions like `torch.cuda.get_device_capability(x.device)[0] >= 9`

File: python/sglang/srt/models/gpt_oss.py (hunk 1 above)
Comment (Qiaolin-Yu) [reply python/sglang/srt/models/gpt_oss.py]: then maybe rename it to `TinyGemmLinear` is better

File: python/sglang/srt/models/gpt_oss.py (hunk 3 above)
Comment (Qiaolin-Yu) [inline python/sglang/srt/models/gpt_oss.py]: `x.shape[0] <= 128`, for decoding, bs 128 will still use this path, right? could you also add decoding performance benchmark results of bs 64, bs 128? Thanks!

Comment (mmangkad) [reply python/sglang/srt/models/gpt_oss.py]: Yes, I’ll follow up with `bs=64/128` decode results

**Review by Qiaolin-Yu — APPROVED** [review]
lgtm. let's wait for CIs.
